# Propensity Score Matching (PSM)

In many medical studies attempting to find out if a particular treatment (tx) had an effect or not, it is not possible to randomly assign subjects into control and experimental groups. This is because generally these studies look at observational data (data over which the experimenter has no control). Properly controlled experimental data can sometimes be difficult to come by because randomly assigning subjects to control and experimental conditions is often unfeasable and/or unethical. 

This leaves us with observational data that is often biased. The group that receives the tx may tend to be younger, sicker, at an earlier or later stage in the disease, etc. In short, apparent differences in outcome between two groups may depend on characteristics that affected whether or not subjects in one group received a given tx instead of due to the effect of the tx per se. PSM attempts to reduce this bias by making the treated and control groups more comparable by matching subjects that are similar (or have a similar propensity for receiving tx) in each group. To do this matching, we can use the k-Nearest Neighbors Algorithm (k-NNA).

We will perform PSM on a depression tx dataset from a NIH study made publically available by UFHealth Biostatistics Open Learning Textbook (http://bolt.mph.ufl.edu/2012/08/02/learn-by-doing-exploring-a-dataset/). ***This dataset has been artificially biased for the purposes of this notebook, the original data has been altered.*** Once we unbias the data using PSM, the original result will be consistent with the results of the same analysis of the original data, but the actualy data will be slightly different. 

Recurrence will be the outcome of interest. Run the two cells below to view the data.

In [20]:
import numpy as np
import pandas as pd
from sklearn.neighbors import NearestNeighbors
from sklearn.preprocessing import normalize
import scipy.stats
from sklearn.linear_model import LogisticRegression


In [21]:
# Data import and preprocessing

data = pd.read_csv('https://raw.githubusercontent.com/BeaverWorksMedlytics/Week1_public/master/Propensity%20Score%20Matching/Depression_data.csv')
data = data.drop('Unnamed: 0', axis = 1)
# Hospt: which hospital patient was at
# Treat: tx recieved
# Outcome: Whether or not recurrence occured during tx
# Time: Days untill recurrence, or if no recurrence, days until observation ended # Will not be using
# AcuteT: Days patient was depressed prior to observation
# Age: age in years
# Gender: 1 = Female, 2 = Male
data

Hospt       Treat        Outcome  AcuteT  Age  Gender
0        1     Lithium     Recurrence     207   33       1
1        1  Imipramine  No Recurrence     217   49       1
2        1  Imipramine  No Recurrence     219   50       1
3        1     Lithium     Recurrence     214   29       2
4        1     Lithium  No Recurrence      63   29       1
5        1     Placebo  No Recurrence      29   30       2
6        1     Lithium  No Recurrence      55   56       1
7        1     Placebo  No Recurrence      97   48       1
8        1     Placebo  No Recurrence     162   22       2
9        1     Placebo  No Recurrence     105   61       2
10       1  Imipramine  No Recurrence     231   58       1
11       1  Imipramine  No Recurrence     200   31       1
12       1  Imipramine  No Recurrence     229   44       1
13       1  Imipramine     Recurrence     236   55       2
14       1     Lithium  No Recurrence     197   57       2
15       1     Lithium     Recurrence     237   46       1
16       1  Imipramine  No Recurrence     203   58       1
17       1     Lithium  No Recurrence     194   57       1
18       2     Lithium     Recurrence     195   54       1
19       2     Lithium  No Recurrence      48   23       1
20       2  Imipramine  No Recurrence     246   65       1
21       2  Imipramine     Recurrence     190   27       1
22       2     Lithium     Recurrence     209   50       1
23       2     Lithium     Recurrence     249   41       2
24       2     Placebo  No Recurrence     129   65       1
25       2     Placebo     Recurrence     208   52       1
26       2     Placebo  No Recurrence     200   60       1
27       2     Placebo     Recurrence     241   25       2
28       2  Imipramine     Recurrence     237   27       2
29       2  Imipramine  No Recurrence     236   32       1
..     ...         ...            ...     ...  ...     ...
79       5     Lithium     Recurrence     231   46       1
80       5  Imipramine  No Recurrence     215   33       2
81       5     Placebo  No Recurrence     476   22       1
82       5     Lithium  No Recurrence     441   37       2
83       6     Lithium     Recurrence     200   40       2
84       6     Lithium  No Recurrence     201   22       1
85       6  Imipramine     Recurrence     200   30       2
86       6     Lithium     Recurrence     235   63       2
87       6  Imipramine  No Recurrence     219   40       1
88       6     Lithium     Recurrence     202   23       1
89       6  Imipramine     Recurrence     240   52       1
90       6     Placebo     Recurrence     240   23       1
91       6     Placebo     Recurrence     216   22       2
92       6     Placebo     Recurrence     231   23       1
93       6     Placebo  No Recurrence     139   30       2
94       6  Imipramine  No Recurrence     211   26       2
95       6     Lithium     Recurrence     212   46       1
96       6     Placebo     Recurrence     242   61       1
97       6  Imipramine     Recurrence     247   30       2
98       6  Imipramine  No Recurrence     222   39       1
99       6  Imipramine  No Recurrence     216   49       1
100      6     Lithium     Recurrence     199   57       1
101      6     Lithium     Recurrence     217   51       1
102      6     Lithium     Recurrence     230   53       1
103      6     Placebo     Recurrence     245   31       1
104      6     Placebo     Recurrence     242   43       1
105      6     Placebo  No Recurrence      84   29       1
106      6  Imipramine     Recurrence     207   28       1
107      6  Imipramine     Recurrence     232   51       1
108      6     Lithium     Recurrence     204   33       1

[109 rows x 6 columns]

# Initial Analysis

Here we will analyse the data without PSM. First we will see if any of the potentially confounding variables have any effect on which tx group each person was placed into. Then we will use a chi square test (a statistical test to determine significant differences in data that deal with frequencies) to see if those treated with lithium or Imipramine have lower recurrence than those who are treated with a placebo. In this analysis we will not take into account any other variables. 





## Pearson's Correlation Coefficient

To see if any of the potentially confounding variables correlate with tx group placement, we can use Pearson's Correlation Coefficient. This is a statistical test that tells you whether two sets of things are correlated with each other. Run this test using:

```
scipy.stats.pearsonr(array, array)
```

Documentation: https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.stats.pearsonr.html

Both arguments shouls be np.arrays and should not contain strings (you may have to recode your Treatment groups as ints instead of strings). Use pearsons's on age, gender, hospt, and acuteT, each compared (seperately) to tx group. This function outputs the correlation coefficient and the p-value (probability of an uncorrelated system producing datasets that have a Pearson correlation at least as extreme as the one computed from these datasets). A correlation coefficient is a number between -1 and +1 that tells you the strength of a correlation. 

![alt test](http://pradeepshettys.files.wordpress.com/2011/09/correlation_coefficient.gif)

In the cell below calculate the correlation coefficient and p-value of tx group with each of the 4 potentially confounding variables. ***You will need to run 8 tests in all. Four comparing between Placebo and Imipramine tx groups and four comparing between Placebo and Lithium tx groups***



In [4]:
placebo_lithium = data[data.Treat != 'Imipramine']  # Df containing only Placebo and Lithium rows
placebo_imipramine = data[data.Treat != 'Lithium']  # Df containing only Placebo and Imipramine rows

age_lith = scipy.stats.pearsonr(np.array(placebo_lithium.Treat.replace(['Placebo', 'Lithium', 'Imipramine'], [0,1,2])), np.array(placebo_lithium.Age))
gender_lith = scipy.stats.pearsonr(np.array(placebo_lithium.Treat.replace(['Placebo', 'Lithium', 'Imipramine'], [0,1,2])), np.array(placebo_lithium.Gender))
hospt_lith = scipy.stats.pearsonr(np.array(placebo_lithium.Treat.replace(['Placebo', 'Lithium', 'Imipramine'], [0,1,2])), np.array(placebo_lithium.Hospt))
AcuteT_lith = scipy.stats.pearsonr(np.array(placebo_lithium.Treat.replace(['Placebo', 'Lithium', 'Imipramine'], [0,1,2])), np.array(placebo_lithium.AcuteT))

age_imip = scipy.stats.pearsonr(np.array(placebo_imipramine.Treat.replace(['Placebo', 'Lithium', 'Imipramine'], [0,1,2])), np.array(placebo_imipramine.Age))
gender_imip = scipy.stats.pearsonr(np.array(placebo_imipramine.Treat.replace(['Placebo', 'Lithium', 'Imipramine'], [0,1,2])), np.array(placebo_imipramine.Gender))
hospt_imip = scipy.stats.pearsonr(np.array(placebo_imipramine.Treat.replace(['Placebo', 'Lithium', 'Imipramine'], [0,1,2])), np.array(placebo_imipramine.Hospt))
AcuteT_imip = scipy.stats.pearsonr(np.array(placebo_imipramine.Treat.replace(['Placebo', 'Lithium', 'Imipramine'], [0,1,2])), np.array(placebo_imipramine.AcuteT))

print('Age_Lith_Plac: ' + str(age_lith),
     '\nGender_Lith_Plac: ' + str(gender_lith),
     '\nHospt_Lith_Plac: ' +str(hospt_lith),
     '\nAcuteT_Lith_Plac: ' + str(AcuteT_lith),
     '\n\nAge_Imip_Plac: ' + str(age_imip),
     '\nGender_Imip_Plac: ' + str(gender_imip),
     '\nHospt_Imip_Plac: ' + str(hospt_imip),
     '\nAcuteT_Imip_Plac: ' + str(AcuteT_imip))

Age_Lith_Plac: (0.13347958216568637, 0.2671154947789644) 
Gender_Lith_Plac: (-0.06069230328831319, 0.6151126599857253) 
Hospt_Lith_Plac: (-0.06069909048400425, 0.6150730517967733) 
AcuteT_Lith_Plac: (0.02028473152924744, 0.8666575780678742) 

Age_Imip_Plac: (0.1325051952046466, 0.2671855579829328) 
Gender_Imip_Plac: (-0.04183179465772309, 0.72716805947342) 
Hospt_Imip_Plac: (-0.06635765932056066, 0.5797020528339649) 
AcuteT_Imip_Plac: (0.2174983804057164, 0.06646523503199259)


You should have gotten:

Age_Lith_Plac: (0.13347958216568637, 0.2671154947789644) 

Gender_Lith_Plac: (-0.06069230328831319, 0.6151126599857253) 

Hospt_Lith_Plac: (-0.06069909048400425, 0.6150730517967733) 

AcuteT_Lith_Plac: (0.02028473152924744, 0.8666575780678742) 


Age_Imip_Plac: (0.1325051952046466, 0.2671855579829328) 

Gender_Imip_Plac: (-0.04183179465772309, 0.72716805947342) 

Hospt_Imip_Plac: (-0.06635765932056066, 0.5797020528339649) 

AcuteT_Imip_Plac: (0.2174983804057164, 0.06646523503199259)



It looks like although none of these tests were statistically significant (AcuteT_Imip_Plac comes close), based on correlation coefficient there is a weak but likely non-zero relationship betwee AcuteT and group tx between the Placebo and Imipramine groups. 


## Contingency Table and Chi Square


First we should look at the frequency of recurrence for each group (lithium, imipramine, and placebo). Put the recurrence and non-reccurence raw frequencies for each tx into a pandas dataframe (columns = Imipramine, Lithium, Placebo; rows = Recurrence, No Recurrence). This kind of table is called a contingency table and is used to compute the chi square test.

Then we should perform a chi square test to see whether these frequencies are significantly different.To perform a chi square test, use **scipy.stats.chi2_contingency()**. The argument for this function is simply the contingency table containing only the two txs (columns) you wish to compare. Perform 2 chi square tests, one comparing placebo and lithium, and the other comparing imipramine and placebo. The p-value of is the second number in the chi square test's output.


In [5]:
contingency_table = pd.DataFrame({'Lithium': (sum(data[data.Treat == 'Lithium']['Outcome'] == 'Recurrence'), sum(data[data.Treat == 'Lithium']['Outcome'] == 'No Recurrence')),
                                  'Imipramine': (sum(data[data.Treat == 'Imipramine']['Outcome'] == 'Recurrence'), sum(data[data.Treat == 'Imipramine']['Outcome'] == 'No Recurrence')),
                                  'Placebo': (sum(data[data.Treat == 'Placebo']['Outcome'] == 'Recurrence'), sum(data[data.Treat == 'Placebo']['Outcome'] == 'No Recurrence'))})

contingency_table = contingency_table.rename(index={0:'Recurrence',1:'No Recurrence'})

contingency_table

Imipramine  Lithium  Placebo
Recurrence             11       23       18
No Recurrence          27       14       16

In [6]:
imip_plac = scipy.stats.chi2_contingency(contingency_table.drop('Lithium', axis = 1))

lith_plac = scipy.stats.chi2_contingency(contingency_table.drop('Imipramine', axis = 1))

print('p-values: ', '\n\nImipramine-Placebo: ' + str(imip_plac[1]), '\nLithium_Placebo: ' + str(lith_plac[1]))

p-values:  

Imipramine-Placebo: 0.06699739739806405 
Lithium_Placebo: 0.5855500699961194


You should have gotten: 
* Imipramine-Placebo: 0.06699739739806405 
* Lithium_Placebo: 0.5855500699961194



This means that if the null hypotheses are true (that there is NO difference between recurrence frequencies who are being treated with lithium/imipramine and with placebo), there is about an 81.4% chance of observing the Lithium data (or data more extreme) and a 6.7% chance of observing the Imipramine data (or data more extreme). By the conventions of classical statistics we have failed to reject the null hypothesis in both cases (alpha = 0.05)
All this basically means that tx with Lithium likely didn't do much to decrease recurrence of depression, and although our p-value for Imipramine-Placebo approached significance, we cannot claim that outcomes for Imipramine tx and placebo tx are significantly different.

Now lets use propensity score matching to attempt to reduce any bias in the placement of patients into each tx group. We will reanalyse the data after matching and compare it to our results here without matching.

# PSM Method 1 - Pure Nearest Neighbors Matching

Nearest Neighbor matching is exactly what it sounds like. We match each patient in one group with its nearest neighbor in the other group based on some distance measurement that takes into account each potentially confounding variable we have data for. In our case, the variables of interest that may affect the probability of being in a given tx group are age, gender, hospital, and acuteT (Time is not of interest because this is a measurement from after the patients are placed in their tx groups). 

Despite the conceptual simplicity, there are a lot of ways we can go about doing this, and each way has its own implicit assumptions built in. Here are a few things to consider:

* Should there be a maximum distance passed which no match should be made? If so, how do we determine that distance?
* Should we match with or without replacement? Without replacement means that each person can be matched only once, while matching with replacement means that one person could be matched with several other people in the other tx group
* What kind of distance should we use?
* Should we weight our features differently? 

For our purposes here, we'll keep things simple and use matching with replacement and without any maximum distance. Before we match, we should normalize our data and break if up into 3 seperate dataframes, one for each tx group. Additionally, we should drop the "Treat" and "Outcome" columns from each dataframe because these will not be used during matching. 

Here are some helpful hints for this process:
* To drop a column from a pandas df, use: **df.drop(['column name'], axis = 1)**
* If you find that your indices are not in order for each of your dataframes use: **df.reset_index(drop=True)**




In [22]:
# Normalize data
for i in [0, 3, 4, 5]:
  data.iloc[:, i] = data.iloc[:, i] / max(data.iloc[:, i].values)
  

# Seperate into 3 dfs (one for each tx group), re-index if needed, and drop Treat and Outcome columns
data_plac = data[data.Treat == 'Placebo'].reset_index(drop=True).drop(["Treat", "Outcome"], axis = 1)
data_lith = data[data.Treat == 'Lithium'].reset_index(drop=True).drop(["Treat", "Outcome"], axis = 1)
data_imip = data[data.Treat == 'Imipramine'].reset_index(drop=True).drop(["Treat", "Outcome"], axis = 1)

Now we can match our data. We will again use sklearn nearest neighbors functions to compute distance. Fill in the code for a function that takes two dataframes and matches the treated (lithium or imipramine) patients to untreated (placebo) patients.




In [23]:
def matches(df_treated, df_untreated):
  treated = df_treated.values         # turn the dataframes into numpy arrays
  untreated = df_untreated.values
  
  nbrs = NearestNeighbors(n_neighbors = 1).fit(untreated)   # Use the NearestNeighbors() function (the only argument is n_neighbors - ie the number of neighbors you would like to find)
                                                            # and the .fit() function to fit the untreated values to the nearest neighbor model. 
                                                            # This will create a model where all the untreated patients are "plotted".
    
  distances, indices = nbrs.kneighbors(treated)             # Use the .kneighbors() function on your model (nbrs) to find the closest untreated match
                                                            # for each treated patient. This function finds the minimum euclidean distance between each 
                                                            # treated point and each untreated point and returns the distance and index of the matches.
  
  indices = indices.reshape(indices.shape[0])  # reshaping indices
  matched = df_untreated.iloc[indices]         # creating df of matched indices
  
  return(matched)



In [24]:
# Run this code to get dfs containing the matched indices for lithium vs. placebo and imipramine vs. placebo
matched_plac_lith = matches(data_lith, data_plac)  
matched_plac_imip = matches(data_imip, data_plac)
lith_plac_matches = pd.DataFrame({'Placebo': matched_plac_lith.index.values, 'Lithium': range(0, len(data_lith))})
imip_plac_matches = pd.DataFrame({'Placebo': matched_plac_imip.index.values, 'Imipramine': range(0, len(data_imip))})


Explore the matched dfs. After matching, how many unique patients do we have in each group for each comparison (placebo vs. lithium and placebo vs. imipramine)?

In [19]:
imip = len(imip_plac_matches)
plac_im = len(list(set(imip_plac_matches['Placebo'].values)))

lith = len(lith_plac_matches)
plac_li = len(list(set(lith_plac_matches['Placebo'].values)))

print('Imipramine: ' + str(imip), 
     '\nPlacebo_matched_Imip: ' + str(plac_im),
     '\n\nLithium: ' + str(lith),
     '\nPlacebo_matched_Lith: ' + str(plac_li))

Imipramine: 38 
Placebo_matched_Imip: 18 

Lithium: 37 
Placebo_matched_Lith: 22


You should have gotten: 

Imipramine: 38 

Placebo_matched_Imip: 18

Lithium: 37 

Placebo_matched_Lith: 22

It looks like we have lost 14 data points for the untreated group in both comparisons. Data loss in PSM happens when there are some patients in the untreated group that do not match well with any patients in the treated group. Data loss is often a cost of PSM. There are ways to reduce the amount of data you lose, but most of them require sacrificing match quality. 


---
# Re-Analysing

We can now re-analyse our data and see if anything changes as a result of matching. Create a contingency table and use the same chi square analysis on the new data set. You can adapt much of the previous code to do this.

In [33]:
indices_plac_lith = list(set(matched_plac_lith.index.values))
indices_plac_imip = list(set(matched_plac_imip.index.values))

placebo_li = data[data.Treat == 'Placebo'].reset_index(drop=True).iloc[indices_plac_lith]  # df of placebo data that matched with lithium patients
lithium = data[data.Treat == 'Lithium']                                                    # df of all lithium patients (all were matched)
placebo_im = data[data.Treat == 'Placebo'].reset_index(drop=True).iloc[indices_plac_imip]  # df of placebo data that matched with imipramine patients
imipramine = data[data.Treat == 'Imipramine']                                              # df of all imipramine patients (all were matched)


contingency_table = pd.DataFrame({'Lithium': (sum(lithium['Outcome'] == 'Recurrence'), sum(lithium['Outcome'] == 'No Recurrence')),
                                  'Imipramine': (sum(imipramine['Outcome'] == 'Recurrence'), sum(imipramine['Outcome'] == 'No Recurrence')),
                                  'Placebo_li': (sum(placebo_li['Outcome'] == 'Recurrence'), sum(placebo_li['Outcome'] == 'No Recurrence')),
                                  'Placebo_im': (sum(placebo_im['Outcome'] == 'Recurrence'), sum(placebo_im['Outcome'] == 'No Recurrence'))})

contingency_table = contingency_table.rename(index={0:'Recurrence',1:'No Recurrence'})

contingency_table


Imipramine  Lithium  Placebo_im  Placebo_li
Recurrence             11       23          13          12
No Recurrence          27       14           5          10

In [34]:
imip_plac = scipy.stats.chi2_contingency(contingency_table.drop(['Lithium', 'Placebo_li'], axis = 1))

lith_plac = scipy.stats.chi2_contingency(contingency_table.drop(['Imipramine', 'Placebo_im'], axis = 1))

print('p-values: ', '\n\nImipramine-Placebo: ' + str(imip_plac[1]), '\nLithium_Placebo: ' + str(lith_plac[1]))

p-values:  

Imipramine-Placebo: 0.005656260767667514 
Lithium_Placebo: 0.7627312590109018


You should have gotten the following p-values:

Imipramine-Placebo: 0.005656260767667514 

Lithium_Placebo: 0.7627312590109018


These results now give evidence that there is a significant difference in outcome between the group treated with Imipramine and the group treated with placebo. In order to confirm that PSM has unbiased our data, we can run the same Pearson Correlation test on our new data. Run the code provided below to see if the dataset is now unbiased.



In [35]:
placebo_lithium = pd.concat([placebo_li, lithium])  # Df containing only Placebo and Lithium rows
placebo_imipramine = pd.concat([placebo_im, imipramine])  # Df containing only Placebo and Imipramine rows

age_lith = scipy.stats.pearsonr(np.array(placebo_lithium.Treat.replace(['Placebo', 'Lithium', 'Imipramine'], [0,1,2])), np.array(placebo_lithium.Age))
gender_lith = scipy.stats.pearsonr(np.array(placebo_lithium.Treat.replace(['Placebo', 'Lithium', 'Imipramine'], [0,1,2])), np.array(placebo_lithium.Gender))
hospt_lith = scipy.stats.pearsonr(np.array(placebo_lithium.Treat.replace(['Placebo', 'Lithium', 'Imipramine'], [0,1,2])), np.array(placebo_lithium.Hospt))
AcuteT_lith = scipy.stats.pearsonr(np.array(placebo_lithium.Treat.replace(['Placebo', 'Lithium', 'Imipramine'], [0,1,2])), np.array(placebo_lithium.AcuteT))

age_imip = scipy.stats.pearsonr(np.array(placebo_imipramine.Treat.replace(['Placebo', 'Lithium', 'Imipramine'], [0,1,2])), np.array(placebo_imipramine.Age))
gender_imip = scipy.stats.pearsonr(np.array(placebo_imipramine.Treat.replace(['Placebo', 'Lithium', 'Imipramine'], [0,1,2])), np.array(placebo_imipramine.Gender))
hospt_imip = scipy.stats.pearsonr(np.array(placebo_imipramine.Treat.replace(['Placebo', 'Lithium', 'Imipramine'], [0,1,2])), np.array(placebo_imipramine.Hospt))
AcuteT_imip = scipy.stats.pearsonr(np.array(placebo_imipramine.Treat.replace(['Placebo', 'Lithium', 'Imipramine'], [0,1,2])), np.array(placebo_imipramine.AcuteT))

print('Age_Lith_Plac: ' + str(age_lith),
     '\nGender_Lith_Plac: ' + str(gender_lith),
     '\nHospt_Lith_Plac: ' +str(hospt_lith),
     '\nAcuteT_Lith_Plac: ' + str(AcuteT_lith),
     '\n\nAge_Imip_Plac: ' + str(age_imip),
     '\nGender_Imip_Plac: ' + str(gender_imip),
     '\nHospt_Imip_Plac: ' + str(hospt_imip),
     '\nAcuteT_Imip_Plac: ' + str(AcuteT_imip))

Age_Lith_Plac: (0.08274898343454508, 0.533233594590881) 
Gender_Lith_Plac: (-0.04015970135349128, 0.7626564411930109) 
Hospt_Lith_Plac: (-0.011496585378099682, 0.9311322791089929) 
AcuteT_Lith_Plac: (0.02283211934469201, 0.8637152323924715) 

Age_Imip_Plac: (-0.0027498930954075163, 0.9839522635731245) 
Gender_Imip_Plac: (-0.045599037870451276, 0.7385989060007107) 
Hospt_Imip_Plac: (-0.11076022139026026, 0.4164106470821225) 
AcuteT_Imip_Plac: (0.09051580076298506, 0.5070457470263025)


According to these correlation coefficients and p-values, the dataset is much less biased than before. PSM has successfully unbiased this dataset and uncovered a relationship between Imipramine tx and outcome!

# PSM Method 2 - Logistic Regression
The idea behind PSM with logistic regression is to create a model trained on the confounding variables of the participants that outputs the predicted probability of each participant belonging to each treatment group. This way, we don't have to worry about weighing the confounding variable according to its effect on tx group, a logistic regression model will decide which variables are most important and weigh them accordingly. After we get each participant's probability of being placed in each of the tx groups (their propensity scores), we can match the participants based on how similar their probabilities are. 

One important thing to note about using logistic regression for PSM is that your training and test data will be the same. This is because we are not concerned with overfitting and our model does not to be generalizable to other data.


Helpful documentation from sklearn.linear_model for logisitc regressions:
* LogisticRegression(): http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html




In [36]:
# Logistic Regression Model 

logistic_data = data.drop('Outcome', axis = 1)
x_train = logistic_data.drop('Treat', axis = 1)
y_train = logistic_data['Treat']

model = LogisticRegression()

model.fit(x_train, y_train)

predictions = pd.DataFrame(model.predict_proba(x_train))
predictions.columns = ['Placebo', 'Lithium', 'Imipramine']

predictions


Placebo   Lithium  Imipramine
0    0.373254  0.342673    0.284072
1    0.389729  0.357508    0.252763
2    0.391249  0.358200    0.250551
3    0.360530  0.322524    0.316946
4    0.318655  0.358580    0.322766
5    0.294750  0.348753    0.356497
6    0.336164  0.389607    0.274229
7    0.345189  0.374731    0.280080
8    0.335392  0.322300    0.342308
9    0.346289  0.373591    0.280120
10   0.401845  0.364236    0.233918
11   0.369069  0.341633    0.289298
12   0.390215  0.350569    0.259216
13   0.390949  0.346464    0.262587
14   0.377703  0.354843    0.267454
15   0.394763  0.351324    0.253913
16   0.391480  0.368835    0.239686
17   0.387399  0.369285    0.243316
18   0.377485  0.364364    0.258151
19   0.300740  0.351357    0.347903
20   0.404942  0.367147    0.227911
21   0.353916  0.336689    0.309395
22   0.379539  0.358012    0.262449
23   0.375597  0.327740    0.296663
24   0.361495  0.386381    0.252124
25   0.380745  0.360239    0.259016
26   0.383973  0.369765    0.246262
27   0.358488  0.312014    0.329498
28   0.358758  0.314696    0.326546
29   0.374903  0.335266    0.289831
..        ...       ...         ...
79   0.359590  0.344344    0.296066
80   0.329384  0.316503    0.354113
81   0.431815  0.282906    0.285279
82   0.420865  0.287422    0.291713
83   0.320980  0.323698    0.355321
84   0.319745  0.319736    0.360519
85   0.312322  0.312305    0.375373
86   0.354303  0.344891    0.300806
87   0.341484  0.337188    0.321328
88   0.320942  0.320705    0.358354
89   0.359461  0.347324    0.293215
90   0.334737  0.315607    0.349657
91   0.311213  0.301266    0.387521
92   0.331448  0.316825    0.351727
93   0.290404  0.320166    0.389430
94   0.312873  0.306365    0.380762
95   0.343905  0.344871    0.311223
96   0.367667  0.356833    0.275500
97   0.329719  0.306020    0.364261
98   0.341755  0.335651    0.322594
99   0.347901  0.347623    0.304476
100  0.348095  0.359097    0.292808
101  0.349938  0.349697    0.300366
102  0.356504  0.349947    0.293550
103  0.343473  0.323609    0.332918
104  0.352628  0.337160    0.310212
105  0.284367  0.342662    0.372970
106  0.326950  0.325555    0.347495
107  0.355595  0.347441    0.296964
108  0.330059  0.331512    0.338429

[109 rows x 3 columns]

## Nearest Neighbors Back Again

Now that we have these probabilities, we need to figure out how to calculate which participants (from different tx groups) are most similar in their probabilities of being in each tx group (i.e. we need to match their propensity scores). Nearest Neighbors is probably one of the fastest and easiest ways to do this. Since we're treating placebo vs. lithium and placebo vs. imipramine as two separate analyses, we will need perform two nearest neighbors matchings using only the probabilities relevant for each match (ex. when matching placebo and imipramine patients, use only the placebo and imipramine probabilities). 

Again, we need to create 3 dataframe, one for each tx group, containing the propensity scores, then match placebo with lithium patients and placebo with imipramine patients. 



In [37]:
# Split into 3 dfs

prop_plac = predictions.iloc[list(data[data.Treat == 'Placebo'].index.values)].reset_index(drop=True)
prop_lith = predictions.iloc[list(data[data.Treat == 'Lithium'].index.values)].reset_index(drop=True)
prop_imip = predictions.iloc[list(data[data.Treat == 'Imipramine'].index.values)].reset_index(drop=True)

In [38]:
# Nearest Neighbors to Match Propensity Scores

def matches(df_treated, df_untreated):
  treated = df_treated.values         # turn the dataframes into numpy arrays
  untreated = df_untreated.values
  
  nbrs = NearestNeighbors(n_neighbors = 1).fit(untreated)   # Use the NearestNeighbors() function (the only argument is n_neighbors - ie the number of neighbors you would like to find)
                                                            # and the .fit() function to fit the untreated values to the nearest neighbor model. 
                                                            # This will create a model where all the untreated patients are "plotted".
    
  distances, indices = nbrs.kneighbors(treated)             # Use the .kneighbors() function on your model (nbrs) to find the closest untreated match
                                                            # for each treated patient. This function finds the minimum euclidean distance between each 
                                                            # treated point and each untreated point and returns the distance and index of the matches.
  
  indices = indices.reshape(indices.shape[0])  # reshaping indices
  matched = df_untreated.iloc[indices]         # creating df of matched indices
  
  return(matched)

In [39]:
matched_plac_lith1 = matches(prop_lith, prop_plac)  
matched_plac_imip1 = matches(prop_imip, prop_plac)
lith_plac_matches1 = pd.DataFrame({'Placebo': matched_plac_lith1.index.values, 'Lithium': range(0, len(prop_lith))})
imip_plac_matches1 = pd.DataFrame({'Placebo': matched_plac_imip1.index.values, 'Imipramine': range(0, len(prop_imip))})


Explore the matched dataframes and see how many unique patients are in each comparison group, just as we did before.



In [40]:
imip = len(imip_plac_matches1)
plac_im = len(list(set(imip_plac_matches1['Placebo'].values)))

lith = len(lith_plac_matches1)
plac_li = len(list(set(lith_plac_matches1['Placebo'].values)))

print('Imipramine: ' + str(imip), 
     '\nPlacebo_matched_Imip: ' + str(plac_im),
     '\n\nLithium: ' + str(lith),
     '\nPlacebo_matched_Lith: ' + str(plac_li))

Imipramine: 38 
Placebo_matched_Imip: 20 

Lithium: 37 
Placebo_matched_Lith: 18


You should have gotten:

Imipramine: 38 
Placebo_matched_Imip: 20

Lithium: 37 
Placebo_matched_Lith: 18

# Re-Re-Analyzing

We will analyze our data one last time according to our new samples. The results should be pretty similar to the pure nearest neighbors approach, but they won't be exactly the same. 

Again, create a contingency table and perform 2 chi-square analyses.

In [41]:
indices_plac_lith = list(set(matched_plac_lith1.index.values))
indices_plac_imip = list(set(matched_plac_imip1.index.values))

placebo_li = data[data.Treat == 'Placebo'].reset_index(drop=True).iloc[indices_plac_lith]  # df of placebo data that matched with lithium patients
lithium = data[data.Treat == 'Lithium']                                                    # df of all lithium patients (all were matched)
placebo_im = data[data.Treat == 'Placebo'].reset_index(drop=True).iloc[indices_plac_imip]  # df of placebo data that matched with imipramine patients
imipramine = data[data.Treat == 'Imipramine']                                              # df of all imipramine patients (all were matched)


contingency_table = pd.DataFrame({'Lithium': (sum(lithium['Outcome'] == 'Recurrence'), sum(lithium['Outcome'] == 'No Recurrence')),
                                  'Imipramine': (sum(imipramine['Outcome'] == 'Recurrence'), sum(imipramine['Outcome'] == 'No Recurrence')),
                                  'Placebo_li': (sum(placebo_li['Outcome'] == 'Recurrence'), sum(placebo_li['Outcome'] == 'No Recurrence')),
                                  'Placebo_im': (sum(placebo_im['Outcome'] == 'Recurrence'), sum(placebo_im['Outcome'] == 'No Recurrence'))})

contingency_table = contingency_table.rename(index={0:'Recurrence',1:'No Recurrence'})

contingency_table


Imipramine  Lithium  Placebo_im  Placebo_li
Recurrence             11       23          15           9
No Recurrence          27       14           5           9

In [42]:
imip_plac = scipy.stats.chi2_contingency(contingency_table.drop(['Lithium', 'Placebo_li'], axis = 1))

lith_plac = scipy.stats.chi2_contingency(contingency_table.drop(['Imipramine', 'Placebo_im'], axis = 1))

print('p-values: ', '\n\nImipramine-Placebo: ' + str(imip_plac[1]), '\nLithium_Placebo: ' + str(lith_plac[1]))

p-values:  

Imipramine-Placebo: 0.00210975487339223 
Lithium_Placebo: 0.5709127961195964


You should have gotten - 

p-values:  

Imipramine-Placebo: 0.00210975487339223 

Lithium_Placebo: 0.5709127961195964

___ 

This method of PSM gives us a very similar result to the pure nearest neighbors method (as it should).

A final note: PSM should generally not be used unless there is a good reason to suspect bias in the dataset. If PSM is used simply with the hope of shifting the p-value slightly by removing subjects, this is called 'p-hacking', and can result in 'significant' results that are not actually valid. P-hacking harms the validity of scientific analysis and can lead to false conclusions with potentially disastrous consequences (https://www.buzzfeed.com/stephaniemlee/brian-wansink-cornell-p-hacking?utm_term=.repyDLKxL#.xtgL7xZ6x).



## Pros and Cons of PSM

### Pros
* Allows for data randomizaton after the fact. This, in some (not all!) cases, can render biased datasets usable for legitimate analysis.
* Effectively reduces bias due to confounding variables that you collected data on.
* Allows for better, more reliable observational analyses.
* In the Logistic Regression approach, each confounding variable is weighted according to its influence on tx group placement.

### Cons
* In the pure nearest neighbors approach, you are likely treating every variable as if they are equally contributing to the probability of a patient being put into a specific tx group. This means that a variable that has no bearing on propensity score is weighted just as much as one that does.
 * This problem can be resolved by weighing each variable differently. For example, different values of a variable have a drastically different probability of being in a tx group, that variable should be weighted more. 
* Tends to throw out data. Especially outlier data. This means that effective PSM requires a large dataset
* Cannot account for confounding variables that are not included in the analysis (duh, it's not magic!)